File: 395-longest-substring-with-at-least-k-repeating-characters/leochang/395.cpp
class Solution {
public:
	int longestSubstring(string s, int k) {
		vector<int> count(26, 0);
		for(char c : s) count[c-'a']++;
		for(int i = 0; i < s.length(); i++) {
			if (count[s[i] - 'a'] < k && count[s[i] - 'a'] > 0) {
				int ans = 0;
				stringstream ss(s);
				string substr;
				while(getline(ss, substr, s[i])) {
					ans = max(ans, longestSubstring(substr, k));
				}
				return ans;
			}
		}
		return s.length();
	}
};

File: 395-longest-substring-with-at-least-k-repeating-characters/ken/395.py
class Solution(object):
    def longestSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        import re

        validSubstrs = []

        def getCharCnt(s):
            cnt = {}

            for c in s:
                if c not in cnt:
                    cnt[c] = 0
                cnt[c] += 1

            return cnt

        def getInvalidChars(s):
            invalid = []

            cnt = getCharCnt(s)
            for c in cnt:
                if cnt[c] < k:
                    invalid.append(c)

            return invalid

        def getValidSubStr(s):
            invalidChars = getInvalidChars(s)

            if len(invalidChars) == 0:
                validSubstrs.append(s)
                return

            pattern = '|'.join(invalidChars)
            candidates = re.split(pattern, s)

            for candidate in candidates:
                if candidate:
                    getValidSubStr(candidate)

        getValidSubStr(s)

        maxLen = 0
        for subStr in validSubstrs:
            maxLen = max(maxLen, len(subStr))

        return maxLen

a = Solution()

# print a.longestSubstring('ababbc', 2)

print a.longestSubstring("bbaaacbd", 3)
